File: src/stores/environment.ts
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import { invoke } from '@tauri-apps/api/core'
import type { Environment, EnvVariable } from '../types'

export const useEnvironmentStore = defineStore('environment', () => {
  const environments = ref<Environment[]>([])
  const variables = ref<EnvVariable[]>([])  // 当前选中环境的变量

  // 当前激活环境（is_active=1）
  const activeEnv = computed<Environment | null>(
    () => environments.value.find(e => e.is_active === 1) ?? null
  )
  const activeEnvId = computed<number | null>(() => activeEnv.value?.id ?? null)

  // ── 环境 CRUD ──────────────────────────────────────────────
  async function loadEnvironments(projectId: number) {
    environments.value = await invoke<Environment[]>('list_environments', { project_id: projectId })
  }

  async function createEnvironment(projectId: number, name: string, baseUrl?: string | null) {
    const env = await invoke<Environment>('create_environment', {
      project_id: projectId,
      name,
      base_url: baseUrl ?? null,
    })
    environments.value.push(env)
    return env
  }

  async function updateEnvironment(id: number, name: string, baseUrl?: string | null) {
    const updated = await invoke<Environment>('update_environment', {
      id,
      name,
      base_url: baseUrl ?? null,
    })
    const idx = environments.value.findIndex(e => e.id === id)
    if (idx !== -1) environments.value[idx] = updated
    return updated
  }

  async function deleteEnvironment(id: number) {
    await invoke('delete_environment', { id })
    environments.value = environments.value.filter(e => e.id !== id)
    if (variables.value.length > 0 && variables.value[0]?.env_id === id) {
      variables.value = []
    }
  }

  async function activateEnvironment(projectId: number, envId: number) {
    await invoke('activate_environment', { project_id: projectId, env_id: envId })
    // 本地更新 is_active
    environments.value.forEach(e => { e.is_active = e.id === envId ? 1 : 0 })
  }

  async function deactivateEnvironment(projectId: number) {
    await invoke('deactivate_environment', { project_id: projectId })
    environments.value.forEach(e => { e.is_active = 0 })
  }

  // ── 环境变量 CRUD ──────────────────────────────────────────
  async function loadVariables(envId: number) {
    variables.value = await invoke<EnvVariable[]>('list_env_variables', { env_id: envId })
  }

  async function createVariable(envId: number, key: string, value: string, description?: string | null) {
    const v = await invoke<EnvVariable>('create_env_variable', {
      env_id: envId, key, value, description: description ?? null,
    })
    variables.value.push(v)
    return v
  }

  async function updateVariable(id: number, key: string, value: string, description: string | null, enabled: number) {
    const updated = await invoke<EnvVariable>('update_env_variable', {
      id, key, value, description, enabled,
    })
    const idx = variables.value.findIndex(v => v.id === id)
    if (idx !== -1) variables.value[idx] = updated
    return updated
  }

  async function deleteVariable(id: number) {
    await invoke('delete_env_variable', { id })
    variables.value = variables.value.filter(v => v.id !== id)
  }

  return {
    environments,
    variables,
    activeEnv,
    activeEnvId,
    loadEnvironments,
    createEnvironment,
    updateEnvironment,
    deleteEnvironment,
    activateEnvironment,
    deactivateEnvironment,
    loadVariables,
    createVariable,
    updateVariable,
    deleteVariable,
  }
})

File: src-tauri/src/commands/environment.rs
use crate::{
    db::AppDb,
    error::CmdResult,
    types::{EnvVariable, Environment},
};
use tauri::State;

/// 获取项目下环境列表
#[tauri::command]
pub async fn list_environments(
    db: State<'_, AppDb>,
    project_id: i64,
) -> CmdResult<Vec<Environment>> {
    let rows = sqlx::query_as::<_, Environment>(
        "SELECT id, project_id, name, base_url, is_active, created_at FROM environments WHERE project_id=? ORDER BY created_at DESC, id DESC",
    )
    .bind(project_id)
    .fetch_all(&db.0)
    .await?;
    Ok(rows)
}

/// 创建环境
#[tauri::command]
pub async fn create_environment(
    db: State<'_, AppDb>,
    project_id: i64,
    name: String,
    base_url: Option<String>,
) -> CmdResult<Environment> {
    let row = sqlx::query_as::<_, Environment>(
        "INSERT INTO environments (project_id, name, base_url) VALUES (?, ?, ?) RETURNING id, project_id, name, base_url, is_active, created_at",
    )
    .bind(project_id)
    .bind(&name)
    .bind(&base_url)
    .fetch_one(&db.0)
    .await?;
    Ok(row)
}

/// 更新环境
#[tauri::command]
pub async fn update_environment(
    db: State<'_, AppDb>,
    id: i64,
    name: String,
    base_url: Option<String>,
) -> CmdResult<Environment> {
    let row = sqlx::query_as::<_, Environment>(
        "UPDATE environments SET name=?, base_url=? WHERE id=? RETURNING id, project_id, name, base_url, is_active, created_at",
    )
    .bind(&name)
    .bind(&base_url)
    .bind(id)
    .fetch_one(&db.0)
    .await?;
    Ok(row)
}

/// 删除环境
#[tauri::command]
pub async fn delete_environment(db: State<'_, AppDb>, id: i64) -> CmdResult<()> {
    sqlx::query("DELETE FROM environments WHERE id=?")
        .bind(id)
        .execute(&db.0)
        .await?;
    Ok(())
}

/// 激活指定环境（先清空项目下激活状态，事务保证原子性）
#[tauri::command]
pub async fn activate_environment(
    db: State<'_, AppDb>,
    project_id: i64,
    env_id: i64,
) -> CmdResult<()> {
    let mut tx = db.0.begin().await?;

    sqlx::query("UPDATE environments SET is_active=0 WHERE project_id=?")
        .bind(project_id)
        .execute(&mut *tx)
        .await?;

    sqlx::query("UPDATE environments SET is_active=1 WHERE id=? AND project_id=?")
        .bind(env_id)
        .bind(project_id)
        .execute(&mut *tx)
        .await?;

    tx.commit().await?;
    Ok(())
}

/// 取消项目下全部环境激活状态
#[tauri::command]
pub async fn deactivate_environment(db: State<'_, AppDb>, project_id: i64) -> CmdResult<()> {
    sqlx::query("UPDATE environments SET is_active=0 WHERE project_id=?")
        .bind(project_id)
        .execute(&db.0)
        .await?;
    Ok(())
}

/// 获取环境变量列表
#[tauri::command]
pub async fn list_env_variables(
    db: State<'_, AppDb>,
    env_id: i64,
) -> CmdResult<Vec<EnvVariable>> {
    let rows = sqlx::query_as::<_, EnvVariable>(
        "SELECT id, env_id, key, value, description, enabled FROM env_variables WHERE env_id=? ORDER BY id",
    )
    .bind(env_id)
    .fetch_all(&db.0)
    .await?;
    Ok(rows)
}

/// 创建环境变量
#[tauri::command]
pub async fn create_env_variable(
    db: State<'_, AppDb>,
    env_id: i64,
    key: String,
    value: String,
    description: Option<String>,
) -> CmdResult<EnvVariable> {
    let row = sqlx::query_as::<_, EnvVariable>(
        "INSERT INTO env_variables (env_id, key, value, description) VALUES (?, ?, ?, ?) RETURNING id, env_id, key, value, description, enabled",
    )
    .bind(env_id)
    .bind(&key)
    .bind(&value)
    .bind(&description)
    .fetch_one(&db.0)
    .await?;
    Ok(row)
}

/// 更新环境变量
#[tauri::command]
pub async fn update_env_variable(
    db: State<'_, AppDb>,
    id: i64,
    key: String,
    value: String,
    description: Option<String>,
    enabled: i64,
) -> CmdResult<EnvVariable> {
    let row = sqlx::query_as::<_, EnvVariable>(
        "UPDATE env_variables SET key=?, value=?, description=?, enabled=? WHERE id=? RETURNING id, env_id, key, value, description, enabled",
    )
    .bind(&key)
    .bind(&value)
    .bind(&description)
    .bind(enabled)
    .bind(id)
    .fetch_one(&db.0)
    .await?;
    Ok(row)
}

/// 删除环境变量
#[tauri::command]
pub async fn delete_env_variable(db: State<'_, AppDb>, id: i64) -> CmdResult<()> {
    sqlx::query("DELETE FROM env_variables WHERE id=?")
        .bind(id)
        .execute(&db.0)
        .await?;
    Ok(())
}
